### PR TITLE
strip semgrep-core for smaller binaries

### PR DESCRIPTION
**Pull request by fopina**
I think `semgrep-core` size is enough to consider looking into how to reduce.

I'm not familiar (at all) with ocaml and I wasn't able to find any compilation flags that would be able to produce smaller binaries.

I'm not a UPX fan (size versus unpack CPU performance, plus many caveats),  `strip` however did yield some results.

Linux aarch64 from ~150 to ~112
OSX aarch64 from ~138 to ~122 🤒 

Googled a bit if there was any known impact of stripping ocaml binaries but couldn't find anything.

[This community thread](https://discuss.ocaml.org/t/stripping-binaries/2308/19) actually confirms many people use it.

### CODE REVIEW

**Comment by aryx** [comment]
the problem is that we might not be able to get stack backtrace when people report issues.

File: Makefile
Comment (aryx) [inline Makefile]: this will have no effect; semgrep-core is not in the current directory.
Did you try 'make core'?

**Review by aryx — CHANGES_REQUESTED** [review]

File: Makefile
Comment (fopina) [reply Makefile]: I haven't tried it locally as my current build env is broken (with missing libraries that should not be missing), so I was relying on the workflows to test, as they upload the artifact 😄 

But indeed I read the `ln -s` line backwards! Updated (and I'll try to fix my local env)

Comment (fopina) [reply Makefile]: Now I managed to run `make core` and confirm strip is working however it feels like there should be a better option in this build system (that I'm not familiar) to hook into...

**Review by aryx — CHANGES_REQUESTED** [review]
still does not work ... semgrep-core is not in the current toplevel directory. it's under bin/

**Comment by fopina** [comment]
> still does not work ... semgrep-core is not in the current toplevel directory. it's under bin/

@aryx this time it was tested locally. “semgrep-core” link is created in the current directory in the line before (that existed).
But it’s a link to a link. In macOS running strip on the link works, maybe not in Linux?

File: Makefile
Comment (fopina) [reply Makefile]: @aryx sorry about all the mess, I still don't understand how `ln` seems to run with a different cwd than the command after (since `semgrep-core` does not exist there).

I've updated to use the full target path and also tested it in a PR directly in my fork ✅

Comment (fopinappb) [reply Makefile]: Ciao @aryx do you think this looks good or should I try another approach?

**Comment by fopinappb** [comment]
@aryx any new feedback?

**Comment by mjambon** [comment]
We had a look into it (with @zzeleznick). It looks like running `strip` with the default options is a good thing. We may want to add comments to the makefile explaining what we now know about this topic.

The findings are there's no known drawback of stripping ocaml (native) executables:

* reduces file size by 12% on MacOS, 21% on Linux,
* runs in under 1 s,
* no known security issues,
* no known performance issues

While not a game changer, `strip` looks simple and solid.

I had a quick look into upx as well. Overall, I don't think we should use it for semgrep:

* great compression rate: -77% without stripping, -82% with stripping (x86_64/Linux),
* takes about 10 s or less to compress semgrep which is too much for the typical OCaml developer who recompiles frequently but otherwise fine in a CI context,
* adds a perceptible startup delay of several hundreds of milliseconds to semgrep,
* worked successfully on the one semgrep scan command I tried.

Other avenues include trying to [slim down Docker images](https://www.ardanlabs.com/blog/2020/04/docker-images-part3-going-farther-reduce-image-size.html).

File: Makefile
Comment (mjambon) [inline Makefile:98]: I'd like to add a comment that links to this PR. Something like this would be useful to future maintainers:
```
# Remove all symbols with GNU strip. It saves 12-21% on the executable size
# and it doesn't seem to reduce the functionality or debuggability of OCaml
# executables. See https://github.com/semgrep/semgrep/pull/9471
```

Another thing: I would run `strip bin/semgrep-core` since it's a simpler path that's human-friendly. `strip` follows symlinks and edits the executable in place.

(I might make these changes myself after merging, we'll see)

Comment (fopina) [reply Makefile:98]: Thanks for the feedback, I'll add the comment.
Regarding the simpler path, that was my first (or one of the first) but, while it worked locally, it kept erroring in the GitHub action saying path wasn't found, even though the line before specially used it
But I'll retry.

**Comment by mjambon** [comment]
GHA checks are stuck in "Expected — Waiting for status to be reported". Not sure what's going on. I'd try merging/rebasing against the `develop` branch and push the result.

File: Makefile
Comment (mjambon) [reply Makefile:98]: Not sure what's going on with the GitHub checks. I'll take it from here -> https://github.com/semgrep/semgrep/pull/9765
Thanks for your contribution!

Comment (fopina) [reply Makefile:98]: Do you mind if I continue to it?
There's a certain satisfaction of getting a "PR merged" notification plus commit authorship 😄

Comment (mjambon) [reply Makefile:98]: Merged.

**Comment by mjambon** [comment]
Merged at https://github.com/semgrep/semgrep/pull/9765

File: Makefile
Comment (fopina) [reply Makefile:98]: Lovely

**Comment by mjambon** [comment]
@fopina I made sure you're credited in the changelog (https://github.com/semgrep/semgrep/pull/9765/files).

**Pull request closed without merging**
